### DISCUSSION TITLE
How to tell a C project to link using a C++ compiler?

### DISCUSSION THREAD

**Discussion by 1480c1** [mesonbuild/meson#13185] [Q&A] [answered]
Hi,

Some files to help illustrate what I'm trying to do.

<details>
<summary>a/a.cc</summary>

```c++
#include <iostream>
extern "C" void hello(void) {
    std::cout << "Hello World\n";
}
```

</details>

<details>
<summary>a/a.h</summary>

```c++
#ifndef A_H
#define A_H

#ifdef __cplusplus
extern "C" {
#endif

void hello(void);

#ifdef __cplusplus
}
#endif

#endif
```

</details>

<details>
<summary>a/meson.build</summary>

```py
project(
  'a',
  'cpp',
  version: '0.1',
  default_options: ['warning_level=3', 'cpp_std=c++14'],
)

pkg = import('pkgconfig')

a = library('a', 'a.cc', 'a.h', install: true)
install_headers('a.h')

pkg.generate(a)
```

</details>

<details>
<summary>b/b.c</summary>

```c++
#include "a.h"

int main() {
    hello();
}
```

</details>

<details>
<summary>b/meson.build</summary>

```c++
project('b', 'c', version: '0.1', default_options: ['warning_level=3'])

a = dependency('a', version: '>=0.1')

executable('b', 'b.c', install: true, dependencies: [a])
```

</details>

I'm essentially compiling two projects, One is a C++ project that isn't using meson, but generates a .pc file.
The other is a C project that use meson, and depends on the C++ project, and uses the .pc file generated by the C++ project.

Both projects are not owned/developed/maintained by me, I'm simply just trying to compile them.

"Normally," project `a`'s .pc file contains `Libs.private: -lstdc++`, which would work fine with `CC=gcc LDFLAGS="-static-libgcc -static-libstdc++"`, but that doesn't work with clang+lld+libc++. To be more specific, when clang is passed both `-static-libstdc++` and `-lstdc++`, it will convert `lstdc++` to the appropriate c++ library when passing to lld, and convert `-static-libstdc++` to `-Bstatic -lc++`.

In my case, this results in lld being called with `ld.lld ... -Bdynamic ... -lc++ -Bstatic -lc++ -Bdynamic ...`, which causes lld to try to link both the static and dynamic versions of libc++, which cases a `duplicate symbol` error. Thus, for my compilations, I've been having to remove any mentions of `-lstdc++` from the .pc files that get generated, and have instead resorted to using `clang++` as the linker (like passing `--ld=$CXX` to FFmpeg's configure for example).

For meson, I could not find any equivalent way to set the linker for a `'c'` project to `$CXX` so clang will implicitly choose the right c++ library to link against. I tried a few other things such as modifying the `CC_LD=clang++` and `LD=clang++`, but the first only seemed to result in `clang -fuse-ld=clang++` (which doesn't work), and the second didn't do anything.

Of course, if I had infinite patience and time, I could either modify the meson.build and change `c` to `cpp` or do what was mentioned in <https://github.com/mesonbuild/meson/discussions/11603> and move one of the .c files to .cc. However, doing that for every meson project would be a massive pain. I'm hoping there's a better way to do this.

#### tldr ####

Trying to find a way to link a C meson project to a C++ library without specifying `-lstdc++` in the linking process, as that causes errors.

**Comment by Andarwinux** [comment, marked as answer]
Try to add --driver-mode=g++ to LDFLAGS?

**Reply by 1480c1** [reply]
That seems to work
```bash
cddeg@ccom-laptop CLANG64 /d/b
$ clang -static-libstdc++ b.c -I /local64/include/ -L /local64/lib/ -la
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
ld.lld: error: undefined symbol: __cxa_begin_catch
>>> referenced by C:/msys64/clang64/include/c++/v1/ostream:685
>>>               liba.a(a.cc.obj):(std::__1::basic_ostream<char, std::__1::char_traits<char>>& std::__1::__put_character_sequence[abi:se180100]<char, std::__1::char_traits<char>>(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, char const*, unsigned long long))
>>> referenced by C:/msys64/clang64/include/c++/v1/ostream:330
>>>               liba.a(a.cc.obj):(std::__1::basic_ostream<char, std::__1::char_traits<char>>::sentry::~sentry())
>>> referenced by liba.a(a.cc.obj):(__clang_call_terminate)
...

cddeg@ccom-laptop CLANG64 /d/b
$ clang --driver-mode=g++ -static-libstdc++ b.c -I /local64/include/ -L /local64/lib/ -la
clang: warning: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated [-Wdeprecated]

cddeg@ccom-laptop CLANG64 /d/b
$ $?
bash: 0: command not found

cddeg@ccom-laptop CLANG64 /d/b
$ PATH=/local64/bin:$PATH ldd a.exe
        ntdll.dll => /c/Windows/SYSTEM32/ntdll.dll (0x7ff8c6570000)
        KERNEL32.DLL => /c/Windows/System32/KERNEL32.DLL (0x7ff8c46f0000)
        KERNELBASE.dll => /c/Windows/System32/KERNELBASE.dll (0x7ff8c3b40000)
        ucrtbase.dll => /c/Windows/System32/ucrtbase.dll (0x7ff8c3ef0000)

```

Do you know if clang or LLVM have any documentation online about what `--driver-mode=` actually does? DuckDuckGo and Google isn't bringing up much except a gist that contains the output of `clang --help-hidden` and a few others, but no documentations. (Seems like a lot of the answers for LLVM is to just look at the code within LLVM)

**Reply by Andarwinux** [reply]
This looks undocumented, I stumbled across the relevant definition in [source code](https://github.com/llvm/llvm-project/blob/90501be35b2c4ad314a45634062e0dfe878d8621/clang/lib/Driver/ToolChain.cpp#L325-L340).
